### PR TITLE
fix(runtime/storage): Wrap the block execution under a transaction

### LINKED ISSUE #4109: Westend sync stopped at #20368319 with `Storage root must match that calculated`
## Describe the bug

The westend instance stopped syncing with the error: `Storage root must match that calculated`.

Before the error happens we have a runtime upgrade to version `1010000`

The logs:

````
2024-07-26T07:55:07Z INFO 🔨 finalised block #20367872 (0x80cb24a923ecf0826b5d0cf711fb25d9cb473938cdd287a9f1ee896a1196abc1), round 272, set id 7695 pkg=state

2024-07-26T07:58:20Z INFO 🔄 detected runtime code change, upgrading with block 0xec26e9d9fcc255c842e87cdee87f97947932deb449399764a4d2f09d1cdefe80 from previous code hash 0xfc05f1863c0050bac8cb117c094232576bf63b8315415e95c77d506ef656a960 to new code hash 0x67b635216cfd9d40548f3a0d0c963015fb2b9a5b2dcf5171e3933b964a6f3838... pkg=state

2024-07-26T07:58:22Z CRITICAL target=runtime message=panicked at /build/substrate/frame/executive/src/lib.rs:834:9: Storage root must match that calculated. ext_logging_log_version_1 pkg=runtime module=wazero

2024-07-26T07:58:22Z ERROR requesting max blocks from best block header: while handling workers results: while handling ready block: processing block data with header and body: handling block: failed to execute block 20368319: running runtime function: wasm error: unreachable
````

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

- The problem at block `#20368319` is related to a miss calculation of removed keys, the clear_prefix method is specified with: 

> Any keys created during the current block execution do not count toward the limit

The behavior to ignore keys created during the current block execution is already implemented however it only applies when the runtime starts a transaction and so to fix the problem we should wrap the whole block execution inside a transaction and commit it when the runtime calls the `Root` method when it is finishing the block

## Tests

```sh
N/A
```

## Issues

- Closes https://github.com/ChainSafe/gossamer/issues/4109

### CODE REVIEW

**Review by jimjbrettj — APPROVED** [review]
Amazing find, great work!

File: lib/runtime/storage/trie.go [hunk 1]
```diff
@@ -146,8 +146,13 @@ func (t *TrieState) MustRoot() common.Hash {
 	return hash
 }
 
-// Root returns the trie's root hash
+// Root is executed in the block finalisation
+// when it is wrapping everything and needs to ensure
+// the root hash matches the expected one, in this case
+// we commit the changeset we started in the beginning
 func (t *TrieState) Root() (common.Hash, error) {
+	t.CommitTransaction()
```
Comment (dimartiro) [inline lib/runtime/storage/trie.go:145]: I have some concerns here.
What if the runtime executes:

1.  `ext_storage_start_transaction` 
2. then commit that transaction `ext_storage_commit_transaction` 
3. and then ask for the root `ext_storage_root`

If I'm not wrong this `t.CommitTransaction()` will panic because of this [condition](https://github.com/ChainSafe/gossamer/blob/76741e489e4b7426e6b4e56699e7d6c32e536f6d/lib/runtime/storage/trie.go#L85-L87)

In other words: are we sure that the runtime wont call `ext_storage_root` twice during the same transaction?

Comment (EclesioMeloJunior) [reply lib/runtime/storage/trie.go:145]: The `ext_storage_root ` is only called in the wrap up of the block, and it happens only once in the whole block execution lifetime, added a warn to this function comment also removed calls to this function from other places in the code.

Comment (dimartiro) [reply lib/runtime/storage/trie.go:145]: Thanks to check that! ❤️

**Review by timwu20 — APPROVED** [review]
Amazing work @EclesioMeloJunior!

**Review by dimartiro — APPROVED** [review]
Great finding!
